2021-Firmware-Bootcamp/Core/Src/main.c

```diff
@@ -0,0 +1,201 @@
+  *
+  * This software component is licensed by ST under BSD 3-Clause license,
+  * the "License"; You may not use this file except in compliance with the
+  * License. You may obtain a copy of the License at:
+  *                        opensource.org/licenses/BSD-3-Clause
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+#include "stm32f0xx_hal.h"
+#include "stm32f0xx_hal_spi.h"
+#include <stdint.h>
+#include <stdio.h>
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+  uint8_t rx_buffer_split[2];
+  uint8_t tx_buffer[2];
+  rx_buffer_split[0] = 0;
+  rx_buffer_split[1] = 0;
+  tx_buffer[0] = 1;
+  tx_buffer[1] = 0b1001;
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+  HAL_TIM_Base_Start(&htim1);
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+    /* USER CODE END WHILE */
+	  if(HAL_SPI_Transmit(&hspi1, &tx_buffer[0], sizeof(tx_buffer[0]), 0) == HAL_ERROR) continue;
+	  if(HAL_SPI_TransmitReceive(&hspi1, &tx_buffer[1], &rx_buffer_split[0], sizeof(tx_buffer[1])+sizeof(rx_buffer_split[0]), 0) == HAL_ERROR) continue;
+	  if(HAL_SPI_Receive(&hspi1, &rx_buffer_split[1], 8, 0) == HAL_ERROR) continue;

```

Comment:
You don't have to call Transmit + TransmitReceive + Receive functions here. You only need TransmitReceive function since it fits this application the best. You want to transmit some command to the adc chip and adc return you the data. I would recommend you take a look at the `HAL_SPI_TransmitReceive` function spec sheet. Hint, the size parameter and also the size of the tx/rx array are supposed to be 3.